tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,309 @@
+    var date = new Date()
+
+    if (!calendar){
+      calendar = await getCalendarByYear(date.getFullYear());
+    }
+
+    let mostRecentPeriodDay = getLastPeriodStart(date, calendar);
+    return mostRecentPeriodDay;
+  },
+
+  /**
+   * Get how far the user is into their period as a percentage
+   * @return {Promise} A percentage approximation (meaning in range [0,1]) of how far the user is into their period
+   */
+  GETCycleDonutPercent: async function() {
+    try{
+      let today = new Date();
+      let today_str = getDateString(today);
+      let percent = await AsyncStorage.getItem(Keys.CYCLE_DONUT_PERCENT)
+      percent = percent != null ? JSON.parse(percent) : null;
+
+
+      let calendar = await getCalendarByYear(date.getFullYear());
+
+      if (percent != null && today_str in percent){
+        return percent[today_str];
+      }
+      else{
+        let mostRecentPeriodStart = await this.GETMostRecentPeriodStartDate(calendar);
+        let avgCycleLength = await this.GETAverageCycleLength(calendar);
+        if (mostRecentPeriodStart && avgCycleLength){
+          let daysSincePeriodStart = getDaysDiff(mostRecentPeriodStart, today);
+          let cycleDonutPercent = daysSincePeriodStart / avgCycleLength;
+          this.POSTCycleDonutPercent(cycleDonutPercent);
+          return cycleDonutPercent;
+        }
+        else{
+          //TODO: is 0 the best value to default to? idk
+          return 0;
+        }
+      }
+
+    } catch(e){
+      console.log(e);
+
+    }
+  },
+
+  /** Get the start and length of each period in the given year
+   * @param {number} year The year to retrieve history for
+   * @return {Promise} an object that contains intervals of the user's period (start & length) in that year
+   */
+  GETCycleHistoryByYear: async function(year) {
+    let intervals = []
+    let endOfYear = new Date(year,11,31);
+    let isYearsLastPeriod = true;
+
+    let calendar = await getCalendarByYear(year);
+
+
+
+    let current = endOfYear;
+    try {
+      // Search backwards until date switches to the previous year
+      while(current.getFullYear() === year){
+        let currentSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth()+1, current.getFullYear());
+
+        if (currentSymptoms.flow !== null && currentSymptoms.flow !== FLOW_LEVEL.NONE){
+          let periodEnd = current;
+          let start = await getLastPeriodStart(current);
+          if (isYearsLastPeriod){
+            periodEnd =  await getNextPeriodEnd(start);
+            isYearsLastPeriod = false;
+          }
+
+
+
+          let periodDays = getDaysDiff(periodEnd, start);
+          intervals.push({"start": start, "periodDays": periodDays})
+          var beforeStart = new Date(start.getTime());
+          beforeStart.setDate(beforeStart.getDate() - 1);
+          current = beforeStart;
+        }
+
+        var yesterday = new Date(current.getTime())
+        yesterday.setDate(current.getDate() - 1)
```

Comment:
something minor but these two lines are used quite often throughout - may be beneficial to modularize them into a simple helper to return the date object we need? beyond taking in the current date as a param the helper can also take in a number (refers to how many days we want before or after the current date) so that we can also use the helper beyond just for getting yesterday but also for getting dates like twoDaysLater, tomorrow, twoDaysEarlier, etc as needed in the methods at the top of the file